File: alchemist-incarnation-protelis/src/main/java/it/unibo/alchemist/model/protelis/actions/SendToNeighbor.java
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.protelis.actions;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import it.unibo.alchemist.model.actions.AbstractAction;
import it.unibo.alchemist.protelis.actions.RunProtelisProgram;
import it.unibo.alchemist.protelis.properties.ProtelisDevice;
import it.unibo.alchemist.model.Context;
import it.unibo.alchemist.model.Node;
import it.unibo.alchemist.model.Reaction;
import it.unibo.alchemist.protelis.AlchemistNetworkManager;

import javax.annotation.Nonnull;
import java.util.List;
import java.util.Objects;

/**
 */
public final class SendToNeighbor extends AbstractAction<Object> {

    private static final long serialVersionUID = -8826563176323247613L;
    private final RunProtelisProgram<?> program;
    private final Reaction<Object> reaction;

    /**
     * @param node
     *            the local node
     * @param reaction
     *            the reaction
     * @param program
     *            the reference {@link RunProtelisProgram}
     */
    @SuppressFBWarnings(value = "EI_EXPOSE_REP2", justification = "This is intentional")
    public SendToNeighbor(final Node<Object> node, final Reaction<Object> reaction, final RunProtelisProgram<?> program) {
        super(node);
        this.reaction = Objects.requireNonNull(reaction);
        this.program = Objects.requireNonNull(program);
        declareDependencyTo(program.asMolecule());
    }

    @Override
    public SendToNeighbor cloneAction(final Node<Object> node, final Reaction<Object> reaction) {
        final ProtelisDevice device = node.asPropertyOrNull(ProtelisDevice.class);
        if (device != null) {
            final List<RunProtelisProgram<?>> possibleRefs = device.allProtelisPrograms();
            if (possibleRefs.size() == 1) {
                return new SendToNeighbor(node, this.reaction, possibleRefs.get(0));
            }
            throw new IllegalStateException(
                "There must be one and one only unconfigured " + RunProtelisProgram.class.getSimpleName()
            );
        }
        throw new IllegalStateException(getClass().getSimpleName() + " cannot get cloned on a node with a missing "
                + ProtelisDevice.class.getSimpleName());
    }

    @Override
    public Context getContext() {
        return Context.NEIGHBORHOOD;
    }

    @Override
    public void execute() {
        final AlchemistNetworkManager mgr = getNode().asProperty(ProtelisDevice.class).getNetworkManager(program);
        Objects.requireNonNull(mgr);
        mgr.simulateMessageArrival(reaction.getTau().toDouble());
        program.prepareForComputationalCycle();
    }

    /**
     * @return the {@link RunProtelisProgram} whose data will be sent
     */
    @Nonnull
    public RunProtelisProgram<?> getProtelisProgram() {
        return program;
    }

    @Override
    public String toString() {
        return "broadcast " + program.asMolecule().getName() + " data";
    }
}

File: alchemist-incarnation-protelis/src/main/kotlin/it/unibo/alchemist/model/protelis/actions/SendToNeighbor.kt
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.protelis.actions

import it.unibo.alchemist.model.Context
import it.unibo.alchemist.model.Node
import it.unibo.alchemist.model.Node.Companion.asProperty
import it.unibo.alchemist.model.Reaction
import it.unibo.alchemist.model.actions.AbstractAction
import it.unibo.alchemist.model.protelis.properties.ProtelisDevice
import java.io.Serial
import java.util.Objects

/**
 * @param node
 * the local node
 * @param reaction
 * the reaction
 * @param protelisProgram the [RunProtelisProgram] whose data will be sent
 */
class SendToNeighbor(
    node: Node<Any>,
    reaction: Reaction<Any>,
    val protelisProgram: RunProtelisProgram<*>,
) : AbstractAction<Any>(node) {
    private val reaction: Reaction<Any> = Objects.requireNonNull<Reaction<Any>>(reaction)

    init {
        declareDependencyTo(protelisProgram.asMolecule())
    }

    override fun cloneAction(
        newNode: Node<Any>,
        newReaction: Reaction<Any>,
    ): SendToNeighbor {
        val device: ProtelisDevice<*> = newNode.asProperty()
        val possibleRefs: List<RunProtelisProgram<*>> = device.allProtelisPrograms()
        check(possibleRefs.size == 1) {
            "There must be one and one only unconfigured " + RunProtelisProgram::class.simpleName
        }
        return SendToNeighbor(newNode, this.reaction, possibleRefs[0])
    }

    override fun getContext(): Context = Context.NEIGHBORHOOD

    override fun execute() {
        val protelisDevice = node.asProperty<ProtelisDevice<*>>(ProtelisDevice::class.java)
        val mgr = protelisDevice.getNetworkManager(this.protelisProgram)
        mgr.simulateMessageArrival(reaction.tau.toDouble())
        protelisProgram.prepareForComputationalCycle()
    }

    override fun toString(): String = "broadcast " + protelisProgram.asMolecule().getName() + " data"

    private companion object {
        @Serial
        private val serialVersionUID = -8826563176323247613L
    }
}
